src/arch_mips.py

```diff
@@ -527,138 +531,155 @@ class MipsArch(Arch):
                 lit = AsmLiteral(((args[2].value + 0x8000) & 0xFFFF) - 0x8000)
-                return Instruction("li", [args[0], lit], instr.meta)
+                return AsmInstruction("li", [args[0], lit])
             if instr.mnemonic == "beq" and args[0] == args[1] == Register("zero"):
-                return Instruction("b", [args[2]], instr.meta)
+                return AsmInstruction("b", [args[2]])
             if instr.mnemonic in ["bne", "beq", "beql", "bnel"] and args[1] == Register(
                 "zero"
             ):
                 mn = instr.mnemonic[:3] + "z" + instr.mnemonic[3:]
-                return Instruction(mn, [args[0], args[2]], instr.meta)
+                return AsmInstruction(mn, [args[0], args[2]])
         if len(args) == 2:
             if instr.mnemonic == "beqz" and args[0] == Register("zero"):
-                return Instruction("b", [args[1]], instr.meta)
+                return AsmInstruction("b", [args[1]])
             if instr.mnemonic == "lui" and isinstance(args[1], AsmLiteral):
                 lit = AsmLiteral((args[1].value & 0xFFFF) << 16)
-                return Instruction("li", [args[0], lit], instr.meta)
+                return AsmInstruction("li", [args[0], lit])
             if instr.mnemonic in LENGTH_THREE:
                 return cls.normalize_instruction(
-                    Instruction(instr.mnemonic, [args[0]] + args, instr.meta)
+                    AsmInstruction(instr.mnemonic, [args[0]] + args)
                 )
         if len(args) == 1:
             if instr.mnemonic in LENGTH_TWO:
                 return cls.normalize_instruction(
-                    Instruction(instr.mnemonic, [args[0]] + args, instr.meta)
+                    AsmInstruction(instr.mnemonic, [args[0]] + args)
                 )
         return instr
 
+    @classmethod
+    def parse(
+        cls, mnemonic: str, args: List[Argument], meta: InstructionMeta
+    ) -> Instruction:
+        jump_target: Optional[Union[JumpTarget, Register]] = None
+        function_target: Optional[Union[AsmGlobalSymbol, Register]] = None
+        has_delay_slot = False
+        is_branch_likely = False
+        is_conditional = False
+        is_return = False
+
+        if mnemonic == "jr" and args[0] == Register("ra"):
+            # Return
+            is_return = True
+            has_delay_slot = True
+        elif mnemonic == "jr":
+            # Jump table (switch)
+            assert isinstance(args[0], Register)
+            jump_target = args[0]
+            is_conditional = True
+            has_delay_slot = True
+        elif mnemonic == "jal":
+            # Function call to label
+            assert isinstance(args[0], AsmGlobalSymbol)
+            function_target = args[0]
+            has_delay_slot = True
+        elif mnemonic == "jalr":
+            # Function call to pointer
+            assert isinstance(args[0], Register)
+            function_target = args[0]
+            has_delay_slot = True
```

Comment:
`jalr` supports a two-argument form: `jalr $a, $b` jumps to `$b` and saves pc+8 to `$a` (`$a` is always `$ra` in practice and elided by most disassemblers.) translate.py supports both forms. We should move that rewrite into normalize_instruction, probably normalizing towards the one-argument form.

Reply:
Moved to `normalize_instruction`.
Technically, the previous code had a specific error for > 2 args, but I removed it because we mostly don't perform those kinds of assertions in translate.py?